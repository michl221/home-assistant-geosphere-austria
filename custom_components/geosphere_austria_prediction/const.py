"""Constants for the GeoSphere Austria Prediction integration."""

from __future__ import annotations

from datetime import timedelta
import logging

from homeassistant.components.weather import (
    ATTR_CONDITION_CLOUDY,
    ATTR_CONDITION_FOG,
    ATTR_CONDITION_LIGHTNING,
    ATTR_CONDITION_PARTLYCLOUDY,
    ATTR_CONDITION_POURING,
    ATTR_CONDITION_RAINY,
    ATTR_CONDITION_SNOWY,
    ATTR_CONDITION_SUNNY,
)

DOMAIN = "geosphere_austria_prediction"
LOGGER = logging.getLogger(__package__)
SCAN_INTERVAL = timedelta(minutes=60)

GSA_TO_HA_CONDITION_MAP = {
    1: ATTR_CONDITION_SUNNY,
    2: ATTR_CONDITION_SUNNY,
    3: ATTR_CONDITION_PARTLYCLOUDY,
    4: ATTR_CONDITION_CLOUDY,
    5: ATTR_CONDITION_CLOUDY,
    6: ATTR_CONDITION_FOG,
    7: ATTR_CONDITION_FOG,
    8: ATTR_CONDITION_RAINY,
    9: ATTR_CONDITION_RAINY,
    10: ATTR_CONDITION_POURING,
    11: ATTR_CONDITION_POURING,
    12: ATTR_CONDITION_POURING,
    13: ATTR_CONDITION_POURING,
    14: ATTR_CONDITION_SNOWY,
    15: ATTR_CONDITION_SNOWY,
    16: ATTR_CONDITION_SNOWY,
    17: ATTR_CONDITION_RAINY,
    18: ATTR_CONDITION_RAINY,
    19: ATTR_CONDITION_RAINY,
    20: ATTR_CONDITION_RAINY,
    21: ATTR_CONDITION_RAINY,
    22: ATTR_CONDITION_RAINY,
    23: ATTR_CONDITION_SNOWY,
    24: ATTR_CONDITION_SNOWY,
    25: ATTR_CONDITION_SNOWY,
    26: ATTR_CONDITION_LIGHTNING,
    27: ATTR_CONDITION_LIGHTNING,
    28: ATTR_CONDITION_LIGHTNING,
    29: ATTR_CONDITION_LIGHTNING,
    30: ATTR_CONDITION_LIGHTNING,
    31: ATTR_CONDITION_LIGHTNING,
    32: ATTR_CONDITION_LIGHTNING,
}
