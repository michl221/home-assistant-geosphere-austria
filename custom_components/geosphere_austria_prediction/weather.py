"""Support for GeoSphere Austria Prediction."""

from __future__ import annotations

from datetime import datetime, time
import math

from homeassistant.components.weather import (
    ATTR_FORECAST_CONDITION,
    ATTR_FORECAST_NATIVE_PRECIPITATION,
    ATTR_FORECAST_NATIVE_TEMP,
    ATTR_FORECAST_NATIVE_TEMP_LOW,
    ATTR_FORECAST_NATIVE_WIND_SPEED,
    ATTR_FORECAST_PRESSURE,
    Forecast,
    SingleCoordinatorWeatherEntity,
    WeatherEntityFeature,
)
from homeassistant.const import (
    UnitOfPrecipitationDepth,
    UnitOfPressure,
    UnitOfSpeed,
    UnitOfTemperature,
)
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
from homeassistant.util import dt as dt_util

from .const import DOMAIN, GSA_TO_HA_CONDITION_MAP, LOGGER
from .coordinator import GeoSphereAustriaPredictionConfigEntry
from .models import Forecast as GeoSphereAustriaForecast


async def async_setup_entry(
    hass: HomeAssistant,
    entry: GeoSphereAustriaPredictionConfigEntry,
    async_add_entities: AddConfigEntryEntitiesCallback,
) -> None:
    """Set up GeoSphere Austria Prediction weather entity based on config entry."""
    coordinator = entry.runtime_data
    async_add_entities(
        [GeoSphereAustriaPredictionWeatherEntity(entry=entry, coordinator=coordinator)]
    )


class GeoSphereAustriaPredictionWeatherEntity(
    SingleCoordinatorWeatherEntity[DataUpdateCoordinator[GeoSphereAustriaForecast]]
):
    """Defines GeoSphere Austria weather entity."""

    _attr_has_entity_name = True
    _attr_name = None
    _attr_native_precipitation_unit = UnitOfPrecipitationDepth.MILLIMETERS
    _attr_native_temperature_unit = UnitOfTemperature.CELSIUS
    _attr_native_wind_speed_unit = UnitOfSpeed.METERS_PER_SECOND
    _attr_supported_features = WeatherEntityFeature.FORECAST_HOURLY
    _attr_native_pressure_unit = UnitOfPressure.HPA

    def __init__(
        self,
        *,
        entry: GeoSphereAustriaPredictionConfigEntry,
        coordinator: DataUpdateCoordinator[GeoSphereAustriaForecast],
    ) -> None:
        """Initialize GeoSphereAustria Prediction weather entity."""
        super().__init__(coordinator=coordinator)
        self._attr_unique_id = entry.entry_id

        self._attr_device_info = DeviceInfo(
            entry_type=DeviceEntryType.SERVICE,
            identifiers={(DOMAIN, entry.entry_id)},
            manufacturer="GeoSphere Austria",
            name=entry.title,
        )

    @property
    def condition(self) -> str | None:
        """Return the current weather condition."""
        if not self.coordinator.data.symbol:
            return None
        return GSA_TO_HA_CONDITION_MAP[self.coordinator.data.symbol[0]]

    @property
    def native_temperature(self) -> float | None:
        """Return the platform temperature."""
        if not self.coordinator.data:
            return None
        return self.coordinator.data.temperature[0]

    @property
    def native_wind_speed(self) -> float | None:
        """Return the wind speed."""
        if (
            not self.coordinator.data.windspeed_eastward
            or not self.coordinator.data.windspeed_northward
        ):
            return None
        return math.sqrt(
            math.pow(self.coordinator.data.windspeed_eastward[0], 2)
            + math.pow(self.coordinator.data.windspeed_northward[0], 2)
        )

    @property
    def native_pressure(self) -> float | None:
        """Return the surface presssure."""
        if not self.coordinator.data.surface_pressure:
            return None
        return self.coordinator.data.surface_pressure[0] / 100

    # @property
    # def wind_bearing(self) -> float | str | None:
    #     """Return the wind bearing."""
    #     if not self.coordinator.data.current_weather:
    #         return None
    #     return self.coordinator.data.current_weather.wind_direction

    @callback
    def _async_forecast_hourly(self) -> list[Forecast] | None:
        """Return the daily forecast in native units."""

        if self.coordinator.data is None:
            return None

        forecasts: list[Forecast] = []

        today = dt_util.utcnow()

        hourly = self.coordinator.data
        for index, _datetime in enumerate(hourly.timestamps):
            if _datetime.tzinfo is None:
                _datetime = _datetime.replace(tzinfo=dt_util.UTC)
            if _datetime < today:
                continue

            forecast = Forecast(
                datetime=_datetime.isoformat(),
            )

            if hourly.temperature is not None:
                forecast[ATTR_FORECAST_NATIVE_TEMP] = hourly.temperature[index]
            if hourly.symbol is not None:
                forecast[ATTR_FORECAST_CONDITION] = GSA_TO_HA_CONDITION_MAP.get(
                    hourly.symbol[index]
                )

            if hourly.precipitation_amount is not None:
                forecast[ATTR_FORECAST_NATIVE_PRECIPITATION] = (
                    hourly.precipitation_amount[index]
                )
            if hourly.minimum_temperature is not None:
                forecast[ATTR_FORECAST_NATIVE_TEMP_LOW] = hourly.minimum_temperature[
                    index
                ]

            if (
                hourly.windspeed_eastward is not None
                and hourly.windspeed_northward is not None
            ):
                forecast[ATTR_FORECAST_NATIVE_WIND_SPEED] = math.sqrt(
                    math.pow(hourly.windspeed_eastward[index], 2)
                    + math.pow(hourly.windspeed_northward[index], 2)
                )
            if hourly.surface_pressure is not None:
                forecast[ATTR_FORECAST_PRESSURE] = hourly.surface_pressure[index] / 100

            forecasts.append(forecast)

        return forecasts
