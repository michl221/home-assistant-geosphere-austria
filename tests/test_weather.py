"""Test for the GeoSphere Austria Prediction weather entity."""

from unittest.mock import AsyncMock

import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.weather import (
    DOMAIN as WEATHER_DOMAIN,
    SERVICE_GET_FORECASTS,
)
from homeassistant.const import ATTR_ENTITY_ID
from homeassistant.core import HomeAssistant

from tests.common import MockConfigEntry


@pytest.mark.freeze_time("2025-09-15T15:00:00Z")
async def test_forecast_service(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    mock_geosphere_austria_prediction: AsyncMock,
    snapshot: SnapshotAssertion,
) -> None:
    """Test forecast service."""
    mock_config_entry.add_to_hass(hass)
    await hass.config_entries.async_setup(mock_config_entry.entry_id)
    await hass.async_block_till_done()

    response = await hass.services.async_call(
        WEATHER_DOMAIN,
        SERVICE_GET_FORECASTS,
        {ATTR_ENTITY_ID: "weather.home", "type": "hourly"},
        blocking=True,
        return_response=True,
    )
    assert response == snapshot(name="forecast_hourly")
