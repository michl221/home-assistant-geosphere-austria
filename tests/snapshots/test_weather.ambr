# serializer version: 1
# name: test_forecast_service[forecast_hourly]
  dict({
    'weather.home': dict({
      'forecast': list([
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-15T15:00:00+00:00',
          'precipitation': 0.0,
          'temperature': 24.7,
          'templow': 24.7,
          'wind_speed': 14.4,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-15T16:00:00+00:00',
          'precipitation': 0.0,
          'temperature': 23.9,
          'templow': 23.9,
          'wind_speed': 9.96,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-15T17:00:00+00:00',
          'precipitation': 0.0,
          'temperature': 23.1,
          'templow': 23.1,
          'wind_speed': 6.84,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-15T18:00:00+00:00',
          'precipitation': 0.0,
          'temperature': 21.1,
          'templow': 21.1,
          'wind_speed': 4.83,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-15T19:00:00+00:00',
          'precipitation': 0.0,
          'temperature': 18.7,
          'templow': 18.7,
          'wind_speed': 6.48,
        }),
        dict({
          'condition': 'lightning',
          'datetime': '2025-09-15T20:00:00+00:00',
          'precipitation': 0.5,
          'temperature': 18.6,
          'templow': 18.2,
          'wind_speed': 11.44,
        }),
        dict({
          'condition': 'lightning',
          'datetime': '2025-09-15T21:00:00+00:00',
          'precipitation': 1.25,
          'temperature': 17.8,
          'templow': 17.8,
          'wind_speed': 12.25,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-15T22:00:00+00:00',
          'precipitation': 1.31,
          'temperature': 17.2,
          'templow': 17.2,
          'wind_speed': 4.1,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-15T23:00:00+00:00',
          'precipitation': 1.31,
          'temperature': 17.0,
          'templow': 17.0,
          'wind_speed': 7.99,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T00:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 17.0,
          'templow': 17.0,
          'wind_speed': 7.42,
        }),
        dict({
          'condition': 'partlycloudy',
          'datetime': '2025-09-16T01:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 16.9,
          'templow': 16.8,
          'wind_speed': 9.47,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T02:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 16.3,
          'templow': 16.3,
          'wind_speed': 10.79,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T03:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 15.9,
          'templow': 15.9,
          'wind_speed': 6.64,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T04:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 15.4,
          'templow': 15.4,
          'wind_speed': 7.2,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T05:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 14.9,
          'templow': 14.9,
          'wind_speed': 8.71,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T06:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 15.3,
          'templow': 14.9,
          'wind_speed': 8.76,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T07:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 15.9,
          'templow': 15.5,
          'wind_speed': 11.44,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T08:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 16.8,
          'templow': 15.9,
          'wind_speed': 19.81,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-16T09:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 17.2,
          'templow': 16.7,
          'wind_speed': 20.93,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-16T10:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 17.7,
          'templow': 17.3,
          'wind_speed': 22.1,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-16T11:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 18.2,
          'templow': 17.7,
          'wind_speed': 22.25,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T12:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 18.6,
          'templow': 18.2,
          'wind_speed': 20.67,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T13:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 18.7,
          'templow': 18.6,
          'wind_speed': 21.71,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T14:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 18.7,
          'templow': 18.6,
          'wind_speed': 20.01,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T15:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 18.5,
          'templow': 18.5,
          'wind_speed': 18.84,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-16T16:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 17.9,
          'templow': 17.9,
          'wind_speed': 18.53,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T17:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 17.5,
          'templow': 17.4,
          'wind_speed': 18.36,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T18:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 16.6,
          'templow': 16.6,
          'wind_speed': 17.7,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-16T19:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 15.8,
          'templow': 15.8,
          'wind_speed': 15.86,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-16T20:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 15.1,
          'templow': 15.1,
          'wind_speed': 14.76,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-16T21:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 14.2,
          'templow': 14.2,
          'wind_speed': 11.72,
        }),
        dict({
          'condition': 'partlycloudy',
          'datetime': '2025-09-16T22:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 13.4,
          'templow': 13.4,
          'wind_speed': 11.84,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-16T23:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 13.1,
          'templow': 13.1,
          'wind_speed': 12.92,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T00:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 13.4,
          'templow': 13.1,
          'wind_speed': 9.57,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T01:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 12.8,
          'templow': 12.8,
          'wind_speed': 11.18,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T02:00:00+00:00',
          'precipitation': 1.32,
          'temperature': 13.2,
          'templow': 12.8,
          'wind_speed': 6.64,
        }),
        dict({
          'condition': 'rainy',
          'datetime': '2025-09-17T03:00:00+00:00',
          'precipitation': 1.47,
          'temperature': 12.8,
          'templow': 12.8,
          'wind_speed': 14.41,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T04:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 12.2,
          'templow': 12.2,
          'wind_speed': 10.37,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T05:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 11.5,
          'templow': 11.5,
          'wind_speed': 6.49,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T06:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 11.4,
          'templow': 11.3,
          'wind_speed': 7.99,
        }),
        dict({
          'condition': 'partlycloudy',
          'datetime': '2025-09-17T07:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 12.7,
          'templow': 11.5,
          'wind_speed': 10.32,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T08:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 14.2,
          'templow': 12.9,
          'wind_speed': 11.59,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T09:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 15.1,
          'templow': 14.1,
          'wind_speed': 11.44,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T10:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 17.0,
          'templow': 15.2,
          'wind_speed': 16.56,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T11:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 17.7,
          'templow': 17.0,
          'wind_speed': 17.64,
        }),
        dict({
          'condition': 'partlycloudy',
          'datetime': '2025-09-17T12:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 18.3,
          'templow': 17.7,
          'wind_speed': 17.31,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T13:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 18.7,
          'templow': 18.3,
          'wind_speed': 16.26,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T14:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 18.9,
          'templow': 18.7,
          'wind_speed': 14.8,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T15:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 18.9,
          'templow': 18.9,
          'wind_speed': 12.73,
        }),
        dict({
          'condition': 'cloudy',
          'datetime': '2025-09-17T16:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 18.5,
          'templow': 18.5,
          'wind_speed': 9.83,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-17T17:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 17.7,
          'templow': 17.7,
          'wind_speed': 9.03,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-17T18:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 16.4,
          'templow': 16.4,
          'wind_speed': 6.16,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-17T19:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 14.8,
          'templow': 14.8,
          'wind_speed': 6.19,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-17T20:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 13.8,
          'templow': 13.8,
          'wind_speed': 6.62,
        }),
        dict({
          'condition': 'sunny',
          'datetime': '2025-09-17T21:00:00+00:00',
          'precipitation': 1.51,
          'temperature': 13.1,
          'templow': 13.1,
          'wind_speed': 7.2,
        }),
      ]),
    }),
  })
# ---
